File: packages/r/tests/testthat/test-render.R
test_that("render_pdf_pages renders a valid PDF to PNG bytes", {
  repo_root <- normalizePath(file.path(getwd(), "..", "..", "..", ".."), mustWork = FALSE)
  pdf_path <- file.path(repo_root, "test_documents", "pdf", "tiny.pdf")

  if (!file.exists(pdf_path)) {
    skip(paste("Test PDF not found at", pdf_path))
  }

  pages <- render_pdf_pages(pdf_path)

  expect_true(is.list(pages))
  expect_true(length(pages) >= 1L)
  for (page_bytes in pages) {
    expect_true(is.raw(page_bytes))
    expect_true(length(page_bytes) > 0L)
    # Check PNG magic bytes
    expect_equal(page_bytes[1:8], as.raw(c(0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a)))
  }
})

test_that("render_pdf_page renders a single page to PNG bytes", {
  repo_root <- normalizePath(file.path(getwd(), "..", "..", "..", ".."), mustWork = FALSE)
  pdf_path <- file.path(repo_root, "test_documents", "pdf", "tiny.pdf")

  if (!file.exists(pdf_path)) {
    skip(paste("Test PDF not found at", pdf_path))
  }

  page_bytes <- render_pdf_page(pdf_path, 0L)

  expect_true(is.raw(page_bytes))
  expect_true(length(page_bytes) > 0L)
  # Check PNG magic bytes
  expect_equal(page_bytes[1:8], as.raw(c(0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a)))
})

test_that("render_pdf_pages errors on nonexistent file", {
  expect_error(render_pdf_pages("/nonexistent/path/to/document.pdf"))
})

test_that("render_pdf_page errors on nonexistent file", {
  expect_error(render_pdf_page("/nonexistent/path/to/document.pdf", 0L))
})

File: e2e/python/tests/test_render.py
"""Tests for PDF rendering API."""
from __future__ import annotations

import pytest

from kreuzberg import render_pdf_page, render_pdf_pages

from . import helpers

PNG_MAGIC = b"\x89PNG\r\n\x1a\n"


def test_render_pdf_pages_happy_path() -> None:
    """Rendering all pages of a valid PDF returns a list of PNG-encoded bytes."""
    document_path = helpers.resolve_document("pdf/tiny.pdf")
    if not document_path.exists():
        pytest.skip(f"Missing test document at {document_path}")

    pages = render_pdf_pages(document_path)

    assert isinstance(pages, list)
    assert len(pages) >= 1
    for page_bytes in pages:
        assert isinstance(page_bytes, bytes)
        assert len(page_bytes) > 0
        assert page_bytes[:8] == PNG_MAGIC, "Expected PNG magic bytes"


def test_render_pdf_page_happy_path() -> None:
    """Rendering a single page of a valid PDF returns PNG-encoded bytes."""
    document_path = helpers.resolve_document("pdf/tiny.pdf")
    if not document_path.exists():
        pytest.skip(f"Missing test document at {document_path}")

    page_bytes = render_pdf_page(document_path, 0)

    assert isinstance(page_bytes, bytes)
    assert len(page_bytes) > 0
    assert page_bytes[:8] == PNG_MAGIC, "Expected PNG magic bytes"


def test_render_pdf_pages_nonexistent_file() -> None:
    """Rendering a nonexistent file raises an error."""
    with pytest.raises(Exception):
        render_pdf_pages("/nonexistent/path/to/document.pdf")


def test_render_pdf_page_nonexistent_file() -> None:
    """Rendering a single page from a nonexistent file raises an error."""
    with pytest.raises(Exception):
        render_pdf_page("/nonexistent/path/to/document.pdf", 0)
